### PR TITLE
fix set_detach visual slide issue

### PR DESCRIPTION
**Pull request by shivajiva101**
pass correct parameters when detaching from an entity.
This fixes the visual aspect of a player detaching from an entity, as the move is currently interpolated from 0,0,0 when a player detaches they disappear and slide back to the correct vector 

issue: https://github.com/minetest/minetest/issues/5620

### CODE REVIEW

**Comment by paramat** [comment]
Tested i guess?

**Comment by stujones11** [comment]
parent, rather cannot seem to edit on this tablet :(

**Comment by paramat** [comment]
I'm no expert on attachments so have no idea if this is correct or not.

**Comment by stujones11** [comment]
Me neither but since the next line checks if parent is NULL then I would assume there is a chance it could be.

**Comment by shivajiva101** [comment]
It was built and tested before making the pr using playerplus, specifically the player knockback as it showed the issue up, boat behaviour was checked

**Comment by stujones11** [comment]
I do not understand how this could possibly fix the reported issue, however, if this is the case then I think it would be wise to run the original `setAttachment` code in the case of a NULL parent.

File: src/script/lua_api/l_object.cpp [hunk 1]
```diff
@@ -729,10 +729,11 @@ int ObjectRef::l_set_detach(lua_State *L)
 	if (parent_id)
 		parent = env->getActiveObject(parent_id);
 
-	// Do it
-	co->setAttachment(0, "", v3f(0,0,0), v3f(0,0,0));
-	if (parent != NULL)
+	if (parent != NULL) {
+		// Do it
+		co->setAttachment(parent->getId(), bone, position, rotation);
```
Comment (SmallJoker) [inline src/script/lua_api/l_object.cpp]: Using `0` instead of `parent->getId()` would be more proper.

**Comment by SmallJoker** [comment]
What about https://github.com/minetest/minetest/blob/master/src/script/lua_api/l_object.cpp#L145 ?

File: src/script/lua_api/l_object.cpp (hunk 1 above)
Comment (shivajiva101) [inline src/script/lua_api/l_object.cpp]: Agreed, that is also a lot simpler! Compiled and tested okay

**Comment by shivajiva101** [comment]
@SmallJoker  The issue happens on the detach not the remove visually if that helps at all

**Comment by stujones11** [comment]
> Compiled and tested okay

Does this code only effect player detachments? I am still concerned about the NULL parent case not being handled as before but perhaps you understand this code better than I do.

**Comment by shivajiva101** [comment]
okay I can satisfy you both by initialising position and rotation to 0,0,0 and dropping parent and bone refs like this...
`v3f position = v3f(0, 0, 0);
v3f rotation = v3f(0, 0, 0);
...
co->setAttachment(0, "", position, rotation);`

File: src/script/lua_api/l_object.cpp [hunk 2]
```diff
@@ -722,15 +722,15 @@ int ObjectRef::l_set_detach(lua_State *L)
...
+	v3f position = v3f(0, 0, 0);
+	v3f rotation = v3f(0, 0, 0);
 	co->getAttachment(&parent_id, &bone, &position, &rotation);
 	ServerActiveObject *parent = NULL;
 	if (parent_id)
 		parent = env->getActiveObject(parent_id);
 
 	// Do it
-	co->setAttachment(0, "", v3f(0,0,0), v3f(0,0,0));
+	co->setAttachment(0, "", position, rotation);
```
Comment (stujones11) [inline src/script/lua_api/l_object.cpp]: Can you be sure that position and rotation are still valid vectors after being passed by reference to `getAttachment`?

Comment (shivajiva101) [inline src/script/lua_api/l_object.cpp]: I'm testing the pr code using 2 players and the knockback code from playerplus, alternating punches cause no visual anomalies showing the vectors are valid for detaching for multiple players but short of initialising them if they null after being passed by the reference to getAttachment the answer is no I cannot be sure

File: src/script/lua_api/l_object.cpp [hunk 3]
```diff
@@ -722,15 +722,15 @@ int ObjectRef::l_set_detach(lua_State *L)
 
 	int parent_id = 0;
 	std::string bone = "";
-	v3f position;
-	v3f rotation;
+	v3f position = v3f(0, 0, 0);
```
Comment (nerzhul) [inline src/script/lua_api/l_object.cpp]: useless, default constructor already inits to 0,0,0

File: src/script/lua_api/l_object.cpp (hunk 2 above)
Comment (stujones11) [inline src/script/lua_api/l_object.cpp]: Looking at the rest of the code, I think this is probably safe.

File: src/script/lua_api/l_object.cpp (hunk 3 above)
Comment (stujones11) [inline src/script/lua_api/l_object.cpp]: This is consistent with the rest of that code, changing that should be another PR

Comment (shivajiva101) [inline src/script/lua_api/l_object.cpp]: interesting, I can find at least 12 instances of it being initialised to default in the file

**Comment by stujones11** [comment]
I doesn't look like anyone is going to commit to this yet it does appear to fix a very annoying bug with entity detachment. Perhaps written like this it should cover all bases.
```diff
@@ -726,11 +726,13 @@ int ObjectRef::l_set_detach(lua_State *L)
        v3f rotation;
        co->getAttachment(&parent_id, &bone, &position, &rotation);
        ServerActiveObject *parent = NULL;
-       if (parent_id)
+       if (parent_id) {
                parent = env->getActiveObject(parent_id);
-
+               co->setAttachment(0, "", position, rotation);
+       } else {
+               co->setAttachment(0, "", v3f(0,0,0), v3f(0,0,0));
+       }
        // Do it
-       co->setAttachment(0, "", v3f(0,0,0), v3f(0,0,0));
        if (parent != NULL)
                parent->removeAttachmentChild(co->getId());
        return 0;
```
FWIW I have tested this code with a number of mods known to make heavy use of attachments and have not found problems.

**Comment by paramat** [comment]
Bug or something very similar seems to have re-emerged in #9140
